### PR TITLE
Ensure template exists when creating data stream

### PR DESCRIPTION
**Pull request by martijnvg**
Limit the creation of data streams only for namespaces that have a composable template with a data stream definition.
This way we ensure that mappings/settings have been specified and will be used at data stream creation and data stream rollover.

Relates to #53100

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java [hunk 1]
```diff
@@ -145,4 +147,16 @@ static ClusterState createDataStream(MetadataCreateIndexService metadataCreateIn
...
+    public static void validateTemplateForDataStream(String dataStreamName, Metadata metadata) {
+        final String v2Template = MetadataIndexTemplateService.findV2Template(metadata, dataStreamName, false);
+        if (v2Template == null) {
+            throw new IllegalArgumentException("no matching index template found for data stream [" + dataStreamName + "]");
+        }
+        IndexTemplateV2 indexTemplateV2 = metadata.templatesV2().get(v2Template);
+        if (indexTemplateV2.getDataStreamTemplate() == null) {
+            throw new IllegalArgumentException("matching index template [" + v2Template + "] for data stream [" + dataStreamName  +
+                "] has no data stream template");
+        }
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:159]: Should we enforce that the timestamp field on the create data stream request is the same as the timestamp field in the data stream template? This can be different if a user provided a different timestamp field via the create data stream api.

If we do enforce this then I think that we should remove the timestamp field from the create data stream request and always use the timestamp field provided in the data stream template inside a itv2.

Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:159]: We should at least make specifying the timestamp field non-mandatory. In the future, it should only be necessary to specify if there is no matching template.

Also, the requirement to have the template have a valid mapping for the timestamp field makes it questionable if specifying a timestamp field on the data stream has any real value. Once/if we add the ability to also specify settings and mappings when creating data streams (i.e., create a data stream without any template requirement), it makes more sense to also be able to specify the timestamp field.

I would opt to remove the ability to specify the timestamp field for the first iteration of data streams.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:159]: Thanks Henning, I will remove the timestamp_field parameter from the create data stream request as part of this PR.

**Comment by dakrone** [comment]
@martijnvg so with this change, it's possible that a datastream that was working can stop working (rolling over) if the index template is deleted. Do you think we should add a warning (and/or an error) when deleting an index template? We could enumerate the datastreams and see if deleting the index meant that one or more datastreams would no longer have a matching template, and warn/error if that is the case. Curious what your thoughts are here

**Comment by dakrone** [comment]
> However I think this change should be as a follow up in order to keep this pr small.

Totally, I don't think it's part of this at all, I was thinking of it as a follow-up as well (doesn't have to be you that does it either). Can you open an issue so we don't forget though?

**Comment by martijnvg** [comment]
> Can you open an issue so we don't forget though?

I've opened: #57004

File: server/src/internalClusterTest/java/org/elasticsearch/action/bulk/BulkIntegrationIT.java [hunk 2]
```diff
@@ -218,14 +217,14 @@ public void testDeleteIndexWhileIndexing() throws Exception {
...
-    public void testMixedAutoCreate() {
+    public void testMixedAutoCreate() throws Exception {
         Settings settings = Settings.builder().put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0).build();
 
         PutIndexTemplateV2Action.Request createTemplateRequest = new PutIndexTemplateV2Action.Request("logs-foo");
         createTemplateRequest.indexTemplate(
             new IndexTemplateV2(
                 List.of("logs-foo*"),
-                new Template(settings, null, null),
+                null,
```
Comment (henningandersen) [inline server/src/internalClusterTest/java/org/elasticsearch/action/bulk/BulkIntegrationIT.java:227]: I think this makes `settings` unused? But I am also uncertain of why this change was made? Just cleanup or a necessity. If the latter, I would like to understand why.

File: server/src/internalClusterTest/java/org/elasticsearch/indices/DataStreamIT.java [hunk 3]
```diff
@@ -72,13 +77,19 @@
 
 public class DataStreamIT extends ESIntegTestCase {
 
+    @After
```
Comment (henningandersen) [inline server/src/internalClusterTest/java/org/elasticsearch/indices/DataStreamIT.java:80]: This is fine, but I wonder if we should follow-up with generically removing all v2 templates like we do for regular templates?

File: server/src/test/java/org/elasticsearch/action/admin/indices/rollover/MetadataRolloverServiceTests.java [hunk 4]
```diff
@@ -515,7 +515,10 @@ public void testRolloverClusterState() throws Exception {
 
     public void testRolloverClusterStateForDataStream() throws Exception {
         final DataStream dataStream = DataStreamTests.randomInstance();
+        IndexTemplateV2 template = new IndexTemplateV2(List.of(dataStream.getName() + "*"), null, null, null, null, null,
+            new IndexTemplateV2.DataStreamTemplate("@timestamp"));
         Metadata.Builder builder = Metadata.builder();
+        builder.put("_id", template);
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/action/admin/indices/rollover/MetadataRolloverServiceTests.java:521]: nit: seems like a strange secret choice of `_id`, maybe just use `template`?

**Review by henningandersen — COMMENTED** [review]
Looking good. I added a number of smaller comments.

File: server/src/internalClusterTest/java/org/elasticsearch/indices/DataStreamIT.java (hunk 3 above)
Comment (martijnvg) [reply server/src/internalClusterTest/java/org/elasticsearch/indices/DataStreamIT.java:80]: Yes, I noticed that component templates were not automatically removed after each test. I don't know why. @dakrone Is there a reason why component templates are automatically removed after each test?

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.data_stream/10_basic.yml [hunk 5]
```diff
@@ -62,62 +80,56 @@
...
   - skip:
-      version: " - 7.8.99"
-      reason: "data streams only supported in 7.9+"
+      version: " - 7.99.99"
+      reason:  "mute bwc until backported"
 
   - do:
       catch: bad_request
       indices.create_data_stream:
         name: invalid-data-stream#-name
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply rest-api-spec/src/main/resources/rest-api-spec/test/indices.data_stream/10_basic.yml:89]: The check whether a template with the same exists happens before validating the name of the template. I can check the error description, so that we ensure that this request fails because of an invalid name.

File: server/src/internalClusterTest/java/org/elasticsearch/action/bulk/BulkIntegrationIT.java (hunk 2 above)
Comment (martijnvg) [reply server/src/internalClusterTest/java/org/elasticsearch/action/bulk/BulkIntegrationIT.java:227]: This was just a cleanup. I will remove the settings variable.

(copy paste from `testAutoCreateV1TemplateNoDataStream()` which ensures that a template applied, but that is not needed here, because if the data stream is created then the template has been applied)

File: docs/reference/indices/create-data-stream.asciidoc [hunk 6]
```diff
@@ -23,18 +23,32 @@ addressed directly, data streams are integrated with the
 <<index-lifecycle-management, index lifecycle management (ILM)>> to facilitate
 the management of the time series data contained in their backing indices.
 
+A data stream can only be created if the namespace it targets has a component
+template exists with a `data_stream` definition.
```
[the remark this answers is not in the document]
Comment (danhermann) [reply docs/reference/indices/create-data-stream.asciidoc:27]: Also, are we calling these composable templates now?

File: docs/reference/indices/create-data-stream.asciidoc [hunk 7]
```diff
@@ -71,11 +85,3 @@ Data stream names must meet the following criteria:
 will count towards the 255 limit faster)
 --
 
-[[indices-create-data-stream-api-request-body]]
-==== {api-request-body-title}
-
-`timestamp_field`::
-(Required, string) The name of the timestamp field. This field must be present
-in all documents indexed into the data stream and must be of type
-<<date, `date`>> or <<date_nanos, `date_nanos`>>.
```
Comment (danhermann) [inline docs/reference/indices/create-data-stream.asciidoc:80]: I think that this documentation on `timestamp_field` should be preserved somewhere else, perhaps in the documentation on creating composable templates.

**Review by danhermann — COMMENTED** [review]
I'm a little concerned that we're moving away from an easy-to-use abstraction by requiring a matching template before a data stream can be created or rolled over. In the simplest case, it's a composable template that adds little value by specifying only the name of the timestamp field. I haven't tried it, but I also wonder what would happen on rollover if the `timestamp_field` was changed in the template. I think it would be preferable if that were an immutable property on a data stream.

Also added a few minor notes on some of the documentation changes.

File: docs/reference/indices/create-data-stream.asciidoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (dakrone) [reply docs/reference/indices/create-data-stream.asciidoc:27]: A component template cannot have a data_stream definition, only an index template is, I would recommend changing it to:

 "if the namespace it targets has a composable index template with a `data_stream` definition."

[the remark this answers is not in the document]
Comment (martijnvg) [reply docs/reference/indices/create-data-stream.asciidoc:27]: whoops, I mixed the terms composable and component up in my head... I meant composable.

**Comment by martijnvg** [comment]
>  I think it would be preferable if that were an immutable property on a data stream.

Once a data stream is created from a composable template then that data stream keeps using the timestamp_field that was specified at the time the data stream was created. If the timestamp_field property is changed in a composable template then that change would only apply for new data streams that are created after that change. The data streams already created keep using the same timestamp_field property, also when doing a rollover.

> but I also wonder what would happen on rollover if the timestamp_field was changed in the template.

I think there is a different issue and that has to do with the mapping of the timestamp_field. A rollover only uses the mappings and settings when doing a rollover to create a new backing index. If the mapping of the timestamp_field has changed then this could be problematic. For example the mapping of a timestamp_field may no longer exist for previous created data streams. 

I think in order to prevent this from happening is adding validation to the put composable template api that prevents changing the mapping for a field, if that field is used as timestamp_field in existing data streams. 

We're going to have to add more validation to the put composable api. For example, timestamp field validation that #56815 tries to add or prevent removing a composable template that is used by a data stream (#57004).

**Comment by martijnvg** [comment]
Something else that crossed my mind is if we make the mapping of a timestamp field part of the `timestamp_field` parameter.

So currently someone would need to configure something like this:

```
PUT /_index_template/template_1
{
  "index_patterns": ["logs-*"],
  "template": {
    "mappings": {
      "properties": {
        "@timestamp": {
          "type": "date"
        }
      }
    }
  },
  "data_stream": {
    "timestamp_field": "@timestamp"
  }
}
```

This requires that we check that the `timestamp_field` parameter and the corresponding mapping are valid. However if we would inline the mapping with the `timestamp_field` then this validation is no longer needed and would make #56815 redundant.

```
PUT /_index_template/template_1
{
  "index_patterns": ["logs-*"],
  "data_stream": {
    "timestamp_field": {
        "@timestamp": {
            "mapping": {
                "type": "date"
                // potential other mapping settings   
            }
        }
    }
  }
}
```

When a new data stream is created from a template then it will carry over this timestamp_field config over from the template's data stream definition. This is immutable and rollover and create data stream logic can use that to always set the timestamp_field's mapping to what has been configured in the data steam definition.

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by danhermann** [comment]
> However if we would inline the mapping with the `timestamp_field` then this validation is no longer needed and would make #56815 redundant.

That sounds good to me.

**Comment by danhermann** [comment]
LGTM pending successful CI.
